### DISCUSSION TITLE
[Feature]: "Add to Library"

### DISCUSSION THREAD

**Discussion by nep2ner** [immich-app/immich#1587] [Feature Request]
### Feature detail

It'd be nice to have an "Add to Library" option for photos that others have shared with you. This could be done on a per-photo or per-album basis.

This way, if someone shares a photo with me via Immich shared album, it's easy for me to set that photo to be viewable from my main photo's tab (what I'm calling a "Library" for the sake of this request) without needing to download it and re-upload it to my own user's Immich library. Sometimes a user might want any new photo added to a shared album to be included in their library, and sometimes a user might just want to pick a particular photo out of a shared album to include in their library.

Even if the original owner deletes the photo on their account, it should remain accessible to other users who have saved that photo to their library.

### Platform

Server

**Comment by alextran1502** [comment]
Neat idea, love it!

**Comment by klejejs** [comment]
This is something that I and my family actively use in Google Photos and it would be very useful to have in Immich as well.

**Comment by dolfje** [comment]
This is also necessary to add a photo from a shared album to your own albums? (Without downloading and reuploading)

**Reply by aviv926** [reply]
In fact, when I want to ensure that a photo stays with me if I was shared with it, I create a new album in my user and then from the shared album I add to the new album I created
And that's how the image always stays (I'm not sure if it's a bug, but maybe @alextran1502  will know)

**Reply by aisbergde** [reply]
at least on the Android mobile client this is not possible. Maybe thos was a bug, which you used as a feature? Or does it still work? Does it create a copy in your own library, or does it "just work" with the asset still only in the library or the contributor?

**Comment by axgd-code** [comment]
Is this topic related to : https://github.com/immich-app/immich/discussions/2730 ?

**Comment by frederikb96** [comment]
I guess this feature could be **the one** which is able to address **all the other feature requests about** shared albums and partner **sharing** not being available in the timeline and the smart functions.

The main problem is that it gets complicated quickly when trying to implement those advanced sharing mechanisms...

The idea could be to go for a modular approach:
- Having this "add to library" function as the lowest level, which will be able to:
  - link the asset to once own library
      - optional: a button to unlink from original asset and do a real copy
      - as long as its linked: changes like date/time or geo are obviously synced
      - all functionalities of the asset are available to all parties, so it will be displayed on timeline, on the map, and the face clustering should use it
      - to not make this even more complicated, for now, I guess it is best that the face clustering does not sync between parties, since this gets so complicated quickly
      - Just trying to use the asset as "virtual copy" but simply not storing it two times on disk
      - but still using separate database entries or so for clustering etc.
  - if the linked asset gets deleted by original user, copy it instead to the sharees libraries
- And then have higher level function blocks which use this lower block, e.g. to:
  -  add a whole album to library (basically calling our add to library for each asset in album)
     - related once, about library, or smart functions:
      - #1779
      - #3176
      - #3984
      - #7047
      - #3780
  - or add all assets of partner to own library/map/faces
    - related once, about library, or smart functions:
    - #6339
    - #5089
    
    
I'm not familiar with the code structure, but I think it would be helpful to try and link a single asset to another person's library and be able to operate on it as if it were your own. This might be the first step in being able to address the other requests at some point  :)

**Reply by alextran1502** [reply]
> I'm not familiar with the code structure, but I think it would be helpful to try and link a single asset to another person's library and be able to operate on it as if it were your own. This might be the first step in being able to address the other requests at some point :)

This is what we are planning to do, to split asset entity away from the owner information. So we can have modular approach on different style of sharing in the future

**Reply by lippertmarkus** [reply]
@alextran1502 I would love to help on this feature but would need some help finding the right places to start

**Reply by bo0tzz** [reply]
If you're on Discord, jump into our contributing channel and we'll chat about it.

**Reply by frederiknolte** [reply]
Hey @alextran1502, is this the "Fine grained access controls" feature that is on the roadmap for this year? If not, do you think this could be put on the roadmap? Considering this feature request exists in various forms with double or triple digit votes

**Reply by alextran1502** [reply]
Yep, it is more or less will depends on that mechanism

**Reply by vadim0872** [reply]
> This is what we are planning to do, to split asset entity away from the owner information. So we can have modular approach on different style of sharing in the future

Will it also be possible to implement data deduplication? 
Currently, if you add the same image/video to different people, it will be saved twice on the HDD.

**Reply by fabb** [reply]
Duplicate photo detection should also automatically stack photos that i have in my own lib, and that someone else shared with me in an album. 
I’m often sharing photos via whatsapp with my partner, and they might end up in a shared album shared with me again.

**Reply by fabb** [reply]
> @alextran1502 This is what we are planning to do, to split asset entity away from the owner information. So we can have modular approach on different style of sharing in the future

How would it be represented in the file system? Per-user takeout needs to be possible, for example in case of a breakup or when one user wants to migrate to a different image service. Such a user needs all his/her own photos and all photos that were shared with them.

**Comment by 4f1sh3r** [comment]
I am very looking forward to this as well. It would be amazing to have this!

**Comment by dwrtewd** [comment]
This would be a great feature, especially as more people start sharing servers.

**Comment by IlBuonTommy** [comment]
Problem statement: Many people share photos because eg. family members use different accounts where they upload their photos. After uploading the photos themselves they can create a shared album. The problem is, that these photos will never make ones timeline. Eg. if i want to view my family members’ photos i always need to go to the shared albums one by one.

Proposed solution: Add an option to show shared photos on the timeline. I think it would be a better experience.

Implementation tips: the goal is to show photos from shared albums in the personal timeline, for me having the ability to edit or delete a specific photo from the album is secondary. Even still having those photos available even though the owner has deleted them is not of primary importance to me, of course it would be great to become the owner of the photos from a shared album in case the previous owner decides to delete the photos, or to suggest to the owner to delete or edit a photo (this feature will open up new problems, such as understanding which bin the deleted photo will go to), or when I share an album with someone (only immich users) I can assign specific roles to users (view only, view and edit, view edit and add). The only thing that could be really useful would be to be able to have facial recognition also on shared albums.

**Comment by DeltaTango69** [comment]
It would be important for us that someone takes care of the personal identification and that this data is identical in both accounts. As I understand the possible implementation, both would have to recognize the person and there could also be different results?!

**Comment by lukashass** [comment]
Just want to add my two cents here and say that I would currently prefer an implementation that is as simple as possible, so basically just copies assets into my own library as if I would upload them manually.

My two main reasons for this are:

- once I added assets to my library I expect that no one else can alter them (or any metadata) in any way
- an implementation based on sharing the same file/database-entries seems a lot more error prone

Anyways, looking forward to this feature, regardless of the approach that will be implemented.

**Comment by Hossy** [comment]
+1 on this.  I'm currently going through pictures of my dad who passed away.  Multiple family members have all added their photos to a single shared album which we are trying to narrow down what goes in the video slideshow for which I created a second shared album.  I can't currently add a photo someone else uploaded/added to another album.  If I try from the multi-photo view, I get a message that says it was already added even though it wasn't.

**Comment by charettepa** [comment]
I think the easiest way to implement this would be to "automate" behind the scenes
using the same process that has been suggested to achieve this

currently
you must download the file, then drag and drop or stick it in the upload folder within your user subfolder
it will then get processed as a fresh upload

so in theory the "Add to Library" button
could simply save it directly to your users upload folder
as this value is already known to the system
i get that it is "not as efficient"
but this would be the quick way to implement it
with possible further enhancements in the future

**Comment by markdepalma** [comment]
Just started using Immich and was upset that I could not create a new album from shared album because all of my wife's pictures were unable to be added. Really looking forward to this.

**Comment by Xizicl** [comment]
### Can we first implement a simple function to merge users or inherit other users' assets?
For example, I have three users A, B, and C. Now B is no longer used, but A wants to inherit all of B's ​​assets, including but not limited to photos and albums.
Ideally, this operation will not cause database confusion, will not lose certain information such as photo location information (EXIF), album structure, etc., and will not generate redundant information.
After a successful merge, B should be as if it never existed, as if these photos were originally uploaded and synchronized by A.
Of course, this operation is actually best performed only by administrators.

**Comment by luxychris2009** [comment]
+1 YES, A VERY NEEDED FEATURE!!  please, we need to be able to add to our library the media that we are shared!  or at least a way, on the back end, to transfer media between users!  thank you!

**Comment by yoannwyffels** [comment]
+1 "Add to my library" will be a good idea...Just copy the selected media files to the user upload directory to be processed by the actual import engine will be enough ;)

**Reply by JcMinarro** [reply]
I don't think copy&paste should be the way to go.
It will duplicate the space for something that wouldn't be needed.
Instead, it should keep a reference for the original file, and doesn't mind where it is "physically stored"

**Reply by yoannwyffels** [reply]
Agree on the duplicate, but if the partner decide to delete the file...then...we will need to mark it as deleted in the partner library, not in our library (so at the end, file will not be deleted physcially).

**Reply by axgd-code** [reply]
One solution would be to let the user choose; a photo could have two states:
– remote,
– local.
It would then work like photo display on a smartphone: both are shown, but the user can choose to save one locally.

**Comment by JaydenWSiena** [comment]
Whats the issue with duplicating the photo? Or at least having the option
to “Copy to Library” while this feature is missing?

**Reply by Crocmagnon** [reply]
> Whats the issue with duplicating the photo?

It takes twice the space.

**Reply by fabb** [reply]
Oooooor, use a proper file system that supports deduplication and Copy on Write, for example BTRF or ZFS. Then files can be actually copied and only take up 1x the space.

**Reply by blublub** [reply]
I already tried that - but when I copied all my images and videos onto the ZFS drive it was in the end larger than the same volume with ext4 - that had something to do with how ZFS writes data to the disk, totally weird but I did find an article with that issue that in some cases it doesn'T work as expected and the savong would only later be available but initially the zpool will reach maximum capacity when a large copy job happens

**Comment by vadim0872** [comment]
@alextran1502 
### **Implementation suggestion:** use a symbolic link mechanism or something similar.
The original photos should be stored separately in another place (for example, library/originals, it doesn't matter), and links to the originals are stored in the library/user1 folder. When user2 clicks "add to my library" the link is added to library/user2. 
When someone deletes a photo, it is checked whether someone has a link to this photo, if there are no more links, then you can delete the original.

**In this case, you can effectively store photos without duplicates and unnecessary io operations when users copy photos to their library.**

**Reply by bo0tzz** [reply]
No need to ping Alex. Symlinks are not a magic solution, they come with a bunch of caveats such as not working across different mount points.

**Reply by vadim0872** [reply]
So I also pointed out that you can use a mechanism similar to symlinks.

**Reply by luxychris2009** [reply]
sounds like a pretty effective and efficient solution.  hopefully it will be considered and looked into soon!  thanks for the input!

**Comment by luxychris2009** [comment]
it would also be important to keep the sidecar information with the photos when copying/moving between users.  in addition to "per photo" ... it would be nice to simply copy/move and entire album!  :)  thank you immich team for brainstorming a grand solution for us here!

**Reply by blublub** [reply]
Could we take it one step further and offer a full deduplication? - so that no identical file will be stored more than once?

In setups with multiple users, i.e families, this will drastically improve/reduce storage requirements.

For this all image data can be inside one single folder and the DB tracks who "uses" the file:
 
1st user uploads

2nd user uploads - since the file is already stored Immich adds a 2nd user to the file

Also user 1 can share/hand over the file with user 2 - if any user deletes a file from its own library the file is stored until all users delete the file.

**Reply by DeltaTango69** [reply]
This would be a very big improvement!

**Reply by Sheyne** [reply]
I wonder if storing your photos by some kind of content hash would be a good solution to this?

**Comment by jmvgmr** [comment]
+1

**Comment by CodeBlooded69** [comment]
This is a very much needed feature!

**Comment by rabbitambulance** [comment]
This feature would be really a gamechanger. As it stands, pictures of my son that my wife shared with me don't show up when I search for pictures of my son, which is just bananas to me.

**Reply by VelocityRa** [reply]
Yeah Immich is unusable for families unless everyone uses the same profile. I recommend that for now.

**Reply by profucius** [reply]
> everyone uses the same profile

Not really a workaround many people are willing to use. I have many private photos that don't make sense to include in a family profile. We just need the feature to be implemented.

**Reply by blublub** [reply]
> This feature would be really a gamechanger. As it stands, pictures of my son that my wife shared with me don't show up when I search for pictures of my son, which is just bananas to me.

You either would need on "family user" or multiple copies of one file, for each user one copy - the latter has some serious storage drawbacks.

 As I mentioned earlier in this post, the best solution would be a per image user and asset management coupled with deduplication, so every file is unique:
User A uploads file 1
User B uploads file 1 (identical file)

Immich recognizes it is the same file and adds a second user to it (including all database fields so everyone can tag it differently).
Sharing would be the some, just without a second user re-uploading it - so user A shares a file and a second owner for that file is created.

Files will only be removed from filesystem when the last user deletes it, so "no more owners".

This principle is "relatively" simple if each user starts with "a blank file", i.e. facial recognition, tags etc. are not inherited from user A - sharing people/facial recognition would make it complicated.

I am no technical expert, but I guess hashing would work and the files would need to be identical on a bit per bit level to be treated as a duplicate - however I cannot say how complicated the whole process would be for immich,  I guess it is not that easy ti implement.

**Comment by niklas85** [comment]
+1 From my side. 

Until today I have expected this feature to be existing in immich. So sad, that it hasn't found the way into the code...

Will be veray happy to use it. Family sharing is possible so. 
In my opinion its no problem to have a picture twice on the disk. For families the sharing feature is more the goal then losing disk space.

Please implement! For me this will be an argument to purchase immich!

**Reply by vadim0872** [reply]
data deduplication is very important. Imagine that you want to share a multi-gigabyte video file with 5 accounts, and if there are many such videos. There is too much extra space for nothing, and so much "extra" space will be needed on the backup disk.

**Comment by profucius** [comment]
There must be a way to do this within the database. Scenario: I take photos and videos and they're added to my Immich acct. I want some but not all of them to be visible in my wife's timeline. In each photo (or bulk select in the timeline view) I could select share to account (instead of share as link). Then I select my wife's account, and then they show up in her timeline.

Even better would be assigning co-ownership. This way my wife could delete them. Scenario: I'm the photographer between me and my wife, I might take a burst of photos of our daughter. My wife can be looking at those photos that I have shared to her timeline, she can delete all the ones that didn't turn out good.

There's no need to duplicate photos and do any other shenanigans outside of the database for this to work. We just need someone on the Immich team to take the mantle on this.

**Comment by nate-griff** [comment]
+1 to this feature. Trying to sync photos someone else uploads so that I can search both their photos and mine for photos of us for a project I'm working on. Right now, I just have to download all of their and upload to my account which is doubling the storage which is kind of a pain.

**Comment by fefe79** [comment]
Is anyone in the dev team actually reading and care enough to make this product more useable to be a viable alternative of bigTech? I ask this seriously as this vital feature requested nearly 3 years ago and not implemented yet.
Duplicate of shared pictures are a waste of space and it requires a good and clever solution to architect/engineer on the DB level!

**Reply by fefe79** [reply]
luxychris2009, You think a feature as basic as this a bad feature that after 3 years of users requests has not been implemented actually is show care?

**Reply by luxychris2009** [reply]
Honestly not sure where all the down-votes are coming from.  This entire thread is full of people thinking we need a feature in regard to sharing photos with other users, especially those locally on your server.

**Reply by ruippeixotog** [reply]
People are downvoting because your comments aren't constructive. Most of the other comments are expressing interest in the feature, explaining their use cases for it and even suggesting possible implementations. You're putting in question the team's care for the product, which they are providing _for free_, just because they happen not to prioritize the same features as you.

This is an open source project. If you want to move to a BigTech alternative, go for it and forget about this. If you think the feature is "basic", write an implementation proposal or send a PR. If you're not going for either, just keep lurking in this thread in silence like the rest of us.

**Reply by fefe79** [reply]
@ ruippeixotog and all the downwoters: i got you on my comment style, it is hard, as well as reality, lets see what IS reality. immich is being promoted and marketed as Google photo’s alternative and has no this feature to share properly worked out, that is one of the basic feature of Google photos! Stop promoting it as an alternative and I would think it is right. On the other hand if you think requesting from the actual developers the basic sharing feature to work as should be is not being made after 3 years of user request and feedback (just look this thread) really. That is a CLEAR sign of “care” or “interest” but take if for what you want the fact that this feature is not made despite users request shows the level of care/interest, wether or not I question face to face, actual fact shows what it is, not me, do not blame me because I have the gut to state reality. I am evaluating immich myself and see if it is worth it, but yes, if I see bad signs from developers/community, I believe that project if not change to serve what users actually request after 3 years, and has no any real comment from devs on it of what the plan, well we see.

**Reply by fefe79** [reply]
To any Dev (if you read anything here, please comment):

1. GUI: share button -> internal user/UUID-link (either general or restricted to an email)

2. Each media should have their own hash (I belive this is exist already as immich not reupload the same hashed image, but I am not considering collision issues right now, although important factor). (DB: new row with media-hash)

3. Register media to each user that tries to upload it and create a hard-link on the server for the user if media exist. (DB: new row with user-whois-uploading-media+media-hash; hard-link on the server) 

4. When sharing, just share it based on a UUID of share as should be. (DB: new share row with UUID point to row in point 2.)

5. Modification should work as any COW system works, but taking care of the user-uploading-media+media-hash and handle hardlink as needed. (DB: should follow COW and ACID principles properly)

6. In order to export a shared media be correct the shared link should have an ACL to control download/export functions, although this will still allow manual download by user one by one for any media (anyway)

**Comment by tjakubo** [comment]
I think it'd be perfectly reasonable to start with an "Upload to my library" operation that 1:1 duplicates the photo (and all metadata that is not user-specific) just like if the viewer of the shared asset uploaded it themselves from a file.

Even if we ever get "real" multi-ownership of assets implemented, copying to own library is STILL sometimes very desirable, e.g. if I want to edit the asset, re-share it without passing forward original author info etc. And it is way to get around the original problem if someone doesn't mind the storage duplication, which is also automagically fixed with specific filesystems and backup solutions.

If the new copying action would be clear on what it does, it's completely tangent, would be useful on its own anyway, and waaay cheaper to implement.

**Reply by cdump** [reply]
We could also add metadata to asset, so we'll be able to migrate that to "dedup" solution when and if it will be implemented, something like
```
    metadata: [
      {
        key: 'shared-library-import',
        value: {
          importedAt: new Date().toISOString(),
          sourceAlbumId,
          sourceAssetId: asset.id,
          sourceOwnerId: asset.ownerId,
          version: 1,
        },
      },
    ],
```

I implemented this approach as a PoC, making only changes to the web frontend, but haven't yet fully tested it.
I also think that backend+clients implementation of this approach is quite simple and cheap to implement.

**Comment by NickMonrad** [comment]
Partner Sharing makes this particularly useful with the newer editor. I understand why I shouldn’t be able to edit or delete my partner’s original asset, but currently there’s no straightforward way to create my own editable copy.

Ideally a partner/shared asset would have “Add/Copy to my library”. The original would remain untouched and owned by the partner, while the copy becomes my asset, preserves the original capture metadata where appropriate, and can then be edited or deleted normally.

Right now the workaround is effectively download/save → re-upload → edit, which feels unnecessarily awkward for a family photo library.
